### PR TITLE
[5.3] [WIP] Batch update with a single query

### PR DESCRIPTION
**Pull request by halaei**
Please don't merge. It is just an idea to be considered for possibly more careful and clean implementations.

Sometimes I face situations when it is important to update multiple rows of a table at once, and performance does matter, especially in terms of the number of queries. Here is an attempt to generally solve the problem using CASE WHEN THEN statements:
## Example

``` php
$newSensorData = [
    12 => ['value' => 32, 'observed_at' => '2016-06-30 12:30:01'],
    13 => ['value' => 33, 'observed_at' => '2016-06-30 12:30:05'],
    16 => ['value' => 30, 'observed_at' => '2016-06-30 12:30:05'],
];
$sensors = Sensor::whereIn('id', array_keys($newSensorData))->get();
foreach($sensors as $sensor) {
    //some calculations then save the new values
    $sensor->value = $newSensorData[$sensor->id]['value'];
    $sensor->observed_at = $newSensorData[$sensor->id]['observed_at'];
}
$sensors->update();
```

The resulting SQL statement (assuming that the old value for sensor#12 is 32):

``` sql
UPDATE `sensors` SET
`value` = CASE 
    WHEN `id` = 13 THEN 33
    WHEN `id` = 16 THEN 30
ELSE `value` END,
`observed_at` = CASE
    WHEN `id` = 12 THEN  '2016-06-30 12:30:01'
    WHEN `id` = 13 THEN '2016-06-30 12:30:05'
    WHEN `id` = 16 THEN '2016-06-30 12:30:05'
ELSE `observed_at` END
WHERE id in (12, 13, 16)
```

I don't know how general is the problem, but when someone face it, it will be hard and messy to generate such a query.

What do you think about the functionality and the best way to implement it?

### CODE REVIEW

File: src/Illuminate/Database/Eloquent/Collection.php [hunk 1]
```diff
@@ -351,4 +351,50 @@ public function toBase()
...
+                $columnUpdate .= " WHEN `$keyName` = ? THEN ?";
+                $params[] = $row[0];
+                $params[] = $row[1];
+            }
+            $columnUpdate .= " ELSE `$column` END";
+            $columnUpdates[] = $columnUpdate;
+        }
+
+        $table = $model->newQuery()->getQuery()->getGrammar()->wrapTable($model->getTable());
+        $sql = "UPDATE $table SET ".join(',', $columnUpdates)." WHERE $keyName IN (".join(',', array_fill(0, count($dirties), '?')).')';
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Collection.php]: Are you sure this works on all database drivers?

File: src/Illuminate/Database/Eloquent/Collection.php [hunk 2]
```diff
@@ -351,4 +351,50 @@ public function toBase()
     {
         return new BaseCollection($this->items);
     }
+
+    public function update()
+    {
+        $dirties = [];
+        $dirtyColumns = [];
+        /** @var Model $model */
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Collection.php]: Please don't include inline phpdoc.

File: src/Illuminate/Database/Eloquent/Collection.php [hunk 3]
```diff
@@ -351,4 +351,50 @@ public function toBase()
     {
         return new BaseCollection($this->items);
     }
+
+    public function update()
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Collection.php]: Please add phpdoc to this method.

File: src/Illuminate/Database/Eloquent/Collection.php (hunk 2 above)
Comment (halaei) [inline src/Illuminate/Database/Eloquent/Collection.php]: I will submit a cleaner PR. Please don't worry about style!

File: src/Illuminate/Database/Eloquent/Collection.php (hunk 1 above)
Comment (halaei) [inline src/Illuminate/Database/Eloquent/Collection.php]: I tested MySQL, and checked the documentations on sqlite. Sounds it is an standard statement. But no I am not sure yet.

File: src/Illuminate/Database/Eloquent/Collection.php (hunk 3 above)
Comment (halaei) [inline src/Illuminate/Database/Eloquent/Collection.php]: OK.

**Comment by taylorotwell** [comment]
Probably will not be merged into core. Could be an interesting helper method you release as open source though.

**Comment by halaei** [comment]
Yes, it will be fun to do that :)

**Pull request closed without merging**